`font-variant-alternates: historical-forms` doesn't reflect properly via StylePropertyMap

cdumez opened the pull request:
#### 37f48652e37e505cc21a3a655d024de17fcc02ce
<pre>
`font-variant-alternates: historical-forms` doesn&apos;t reflect properly via StylePropertyMap
<a href="https://bugs.webkit.org/show_bug.cgi?id=249337">https://bugs.webkit.org/show_bug.cgi?id=249337</a>

Reviewed by Tim Nguyen.

`font-variant-alternates: historical-forms` doesn&apos;t reflect properly via
StylePropertyMap. Even though `historical-forms` is a supported keyword for
this CSS property [1], we wouldn&apos;t get a CSSKeywordValue when reading the
property value back. The issue is that we convert to a FontVariantAlternates
internally and then back to a CSSValue. The function to convert back to a
CSSValue (fontVariantAlternatesPropertyValue) was only handling the `normal`
keyword but not the `historical-forms` one.

[1] <a href="https://w3c.github.io/csswg-drafts/css-fonts/#font-variant-alternates-prop">https://w3c.github.io/csswg-drafts/css-fonts/#font-variant-alternates-prop</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/font-variant-alternates-expected.txt:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::fontVariantAlternatesPropertyValue):

Canonical link: <a href="https://commits.webkit.org/257955@main">https://commits.webkit.org/257955@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7df940c0bd6c8386e8aab279e310937d0bdc6219

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100443 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9603 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33505 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/109777 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170026 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104435 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10515 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/182 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/92863 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107632 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106222 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/7955 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91226 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/92863 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89872 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22615 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/92863 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3339 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24134 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3331 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/281 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9452 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43624 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5146 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2848 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -1477,6 +1477,8 @@ static Ref<CSSValue> fontVariantAlternatesPropertyValue(FontVariantAlternates al
     if (alternates.isNormal())
         return CSSValuePool::singleton().createIdentifierValue(CSSValueNormal);
+    if (alternates.values().historicalForms)
+        return CSSValuePool::singleton().createIdentifierValue(CSSValueHistoricalForms);
```
Does this work correctly if you have `font-variant-alternates: historical-forms stylistic(foo)`?

cdumez replied:
Probably not, I'll add coverage and fix as needed. I didn't realize this was valid syntax. However, I just checked the spec and you're right.

cdumez replied:
Fixed.

nt1m reviewed (approved):
This is probably fine, though I question the spec here.

Seems surprising from a web developer perspective that `historical-forms`  and `historical-forms swash(flowing)` return different things that can be manipulated/accessed in completely different ways.

cdumez committed:
`font-variant-alternates: historical-forms` doesn't reflect properly via StylePropertyMap
https://bugs.webkit.org/show_bug.cgi?id=249337

Reviewed by Tim Nguyen.

`font-variant-alternates: historical-forms` doesn't reflect properly via
StylePropertyMap. Even though `historical-forms` is a supported keyword for
this CSS property [1], we wouldn't get a CSSKeywordValue when reading the
property value back. The issue is that we convert to a FontVariantAlternates
internally and then back to a CSSValue. The function to convert back to a
CSSValue (fontVariantAlternatesPropertyValue) was only handling the `normal`
keyword but not the `historical-forms` one.

[1] https://w3c.github.io/csswg-drafts/css-fonts/#font-variant-alternates-prop

* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/font-variant-alternates-expected.txt:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::fontVariantAlternatesPropertyValue):

Canonical link: https://commits.webkit.org/257955@main
